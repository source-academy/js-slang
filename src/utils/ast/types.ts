// Reexport the type definitions from estree, but with several helper types as well

import type * as es from 'estree'

import type { Replace } from '../../types'

export type ExportDeclaration = Exclude<es.ModuleDeclaration, es.ImportDeclaration>
export type ImportSpecifiers =
  | es.ImportSpecifier
  | es.ImportDefaultSpecifier
  | es.ImportNamespaceSpecifier
export type ModuleDeclarationWithSource = Exclude<es.ModuleDeclaration, es.ExportDefaultDeclaration>
export type BlockArrowFunctionExpression = Replace<
  es.ArrowFunctionExpression,
  {
    expression: false
    body: es.BlockStatement
  }
>

export type FunctionNode =
  | es.FunctionDeclaration
  | es.FunctionExpression
  | es.ArrowFunctionExpression

export type FunctionDeclarationWithId = Replace<es.FunctionDeclaration, { id: es.Identifier }>
export type ClassDeclarationWithId = Replace<es.ClassDeclaration, { id: es.Identifier }>

export type ForStatements = es.ForInStatement | es.ForOfStatement | es.ForStatement
export type LoopNode = es.WhileStatement | ForStatements

/**
 * Represents exports of the form `export { a, b } from './a.js';`
 */
export type ExportNamedDeclarationWithSource = Replace<es.ExportNamedDeclaration, {
  declaration: null,
  source: es.Literal
}>

export * from 'estree'
