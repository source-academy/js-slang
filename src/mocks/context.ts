import * as es from 'estree'

import createContext from '../createContext'
import Closure from '../interpreter/closure'
import { Context, Environment, Variant } from '../types'

export function mockContext(chapter = 1, variant: Variant = 'default'): Context {
  return createContext(chapter, variant)
}

export function mockRuntimeContext(): Context {
  const context = createContext()
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ]
  }
  return context
}

export function mockClosure(): Closure {
  return new Closure(
    {
      type: 'FunctionExpression',
      loc: null,
      id: null,
      params: [],
      body: {
        type: 'BlockStatement',
        body: []
      }
    } as es.FunctionExpression,
    {} as Environment,
    {} as Context
  )
}
