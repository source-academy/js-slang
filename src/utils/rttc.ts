import * as es from 'estree'
import { RuntimeSourceError } from '../errors/runtimeSourceError'
import { ErrorSeverity, ErrorType, Value } from '../types'

const LHS = ' on left hand side of operation'
const RHS = ' on right hand side of operation'

export class TypeError extends RuntimeSourceError {
  public type = ErrorType.RUNTIME
  public severity = ErrorSeverity.ERROR
  public location: es.SourceLocation

  constructor(node: es.Node, public side: string, public expected: string, public got: string) {
    super(node)
  }

  public explain() {
    return `Expected ${this.expected}${this.side}, got ${this.got}.`
  }

  public elaborate() {
    return this.explain()
  }
}

// We need to define our own typeof in order for null/array to display properly in error messages
const typeOf = (v: Value) => {
  if (v === null) {
    return 'null'
  } else if (Array.isArray(v)) {
    return 'array'
  } else {
    return typeof v
  }
}

const isNumber = (v: Value) => typeOf(v) === 'number'
// See section 4 of https://2ality.com/2012/12/arrays.html
// v >>> 0 === v checks that v is a valid unsigned 32-bit int
// tslint:disable-next-line:no-bitwise
const isArrayIndex = (v: Value) => isNumber(v) && v >>> 0 === v && v < 2 ** 32 - 1
const isString = (v: Value) => typeOf(v) === 'string'
const isBool = (v: Value) => typeOf(v) === 'boolean'
const isObject = (v: Value) => typeOf(v) === 'object'
const isArray = (v: Value) => typeOf(v) === 'array'

export const checkUnaryExpression = (node: es.Node, operator: es.UnaryOperator, value: Value) => {
  if ((operator === '+' || operator === '-') && !isNumber(value)) {
    return new TypeError(node, '', 'number', typeOf(value))
  } else if (operator === '!' && !isBool(value)) {
    return new TypeError(node, '', 'boolean', typeOf(value))
  } else {
    return undefined
  }
}

export const checkBinaryExpression = (
  node: es.Node,
  operator: es.BinaryOperator,
  left: Value,
  right: Value
) => {
  switch (operator) {
    case '-':
    case '*':
    case '/':
    case '%':
      if (!isNumber(left)) {
        return new TypeError(node, LHS, 'number', typeOf(left))
      } else if (!isNumber(right)) {
        return new TypeError(node, RHS, 'number', typeOf(right))
      } else {
        return
      }
    case '+':
    case '<':
    case '<=':
    case '>':
    case '>=':
      if (isNumber(left)) {
        return isNumber(right) ? undefined : new TypeError(node, RHS, 'number', typeOf(right))
      } else if (isString(left)) {
        return isString(right) ? undefined : new TypeError(node, RHS, 'string', typeOf(right))
      } else {
        return new TypeError(node, LHS, 'string or number', typeOf(left))
      }
    case '!==':
    case '===':
    default:
      return
  }
}

export const checkIfStatement = (node: es.Node, test: Value) => {
  return isBool(test) ? undefined : new TypeError(node, ' as condition', 'boolean', typeOf(test))
}

export const checkMemberAccess = (node: es.Node, obj: Value, prop: Value) => {
  if (isObject(obj)) {
    return isString(prop) ? undefined : new TypeError(node, ' as prop', 'string', typeOf(prop))
  } else if (isArray(obj)) {
    return isArrayIndex(prop)
      ? undefined
      : isNumber(prop)
      ? new TypeError(node, ' as prop', 'array index', 'other number')
      : new TypeError(node, ' as prop', 'array index', typeOf(prop))
  } else {
    return new TypeError(node, '', 'object or array', typeOf(obj))
  }
}

export const isIdentifier = (node: any): node is es.Identifier => {
  return (node as es.Identifier).name !== undefined
}
