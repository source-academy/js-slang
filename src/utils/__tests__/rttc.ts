import { BinaryOperator, UnaryOperator } from 'estree'
import { mockClosure, mockRuntimeContext } from '../../mocks/context'
import { Value } from '../../types'
import * as rttc from '../rttc'

const num = 0
const bool = true
const str = ' '
const func = mockClosure()
const builtin = (x: Value) => x
const obj = { a: 1 }
const arr = [2]

const mockValues: Value[] = [num, bool, str, func, builtin, obj, arr, undefined, null]

describe('Unary type combinations:', () => {
  const valid: Array<[UnaryOperator, Value]> = [['!', bool], ['+', num], ['-', num]]
  const operators: UnaryOperator[] = ['!', '+', '-']
  const invalid: Array<[UnaryOperator, Value]> = []

  operators.forEach(op => {
    mockValues.forEach(value => {
      if (!valid.some(combination => combination[0] === op && combination[1] === value)) {
        invalid.push([op, value])
      }
    })
  })

  test('Valid type combinations are OK', () => {
    valid.forEach(([operator, value]: [UnaryOperator, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkUnaryExpression(context, operator, value)
      expect(error).toBeUndefined()
    })
  })

  test('Invalid type combinations return TypeError', () => {
    invalid.forEach(([operator, value]: [UnaryOperator, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkUnaryExpression(context, operator, value)
      expect(error).toBeInstanceOf(rttc.TypeError)
      expect({
        operator,
        value,
        explain: error!.explain(),
        elaborate: error!.elaborate()
      }).toMatchSnapshot()
    })
  })
})

describe('Binary + type combinations:', () => {
  const valid: Array<[BinaryOperator, Value, Value]> = [['+', num, num], ['+', str, str]]
  const operators: BinaryOperator[] = ['+']
  const invalid: Array<[BinaryOperator, Value, Value]> = []

  operators.forEach(op => {
    mockValues.forEach(left => {
      mockValues.forEach(right => {
        if (
          !valid.some(
            combination =>
              combination[0] === op && combination[1] === left && combination[2] === right
          )
        ) {
          invalid.push([op, left, right])
        }
      })
    })
  })

  test('Valid type combinations are OK', () => {
    valid.forEach(([operator, left, right]: [BinaryOperator, Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkBinaryExpression(context, operator, left, right)
      expect(error).toBeUndefined()
    })
  })

  test('Invalid type combinations return TypeError', () => {
    invalid.forEach(([operator, left, right]: [BinaryOperator, Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkBinaryExpression(context, operator, left, right)
      expect(error).toBeInstanceOf(rttc.TypeError)
      expect({
        operator,
        left,
        right,
        explain: error!.explain(),
        elaborate: error!.elaborate()
      }).toMatchSnapshot()
    })
  })
})

describe('Binary (-|*|/|%) type combinations:', () => {
  const valid: Array<[BinaryOperator, Value, Value]> = [
    ['-', num, num],
    ['*', num, num],
    ['/', num, num],
    ['%', num, num]
  ]
  const operators: BinaryOperator[] = ['-', '*', '/', '%']
  const invalid: Array<[BinaryOperator, Value, Value]> = []

  operators.forEach(op => {
    mockValues.forEach(left => {
      mockValues.forEach(right => {
        if (
          !valid.some(
            combination =>
              combination[0] === op && combination[1] === left && combination[2] === right
          )
        ) {
          invalid.push([op, left, right])
        }
      })
    })
  })

  test('Valid type combinations are OK', () => {
    valid.forEach(([operator, left, right]: [BinaryOperator, Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkBinaryExpression(context, operator, left, right)
      expect(error).toBeUndefined()
    })
  })

  test('Invalid type combinations return TypeError', () => {
    invalid.forEach(([operator, left, right]: [BinaryOperator, Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkBinaryExpression(context, operator, left, right)
      expect(error).toBeInstanceOf(rttc.TypeError)
      expect({
        operator,
        left,
        right,
        explain: error!.explain(),
        elaborate: error!.elaborate()
      }).toMatchSnapshot()
    })
  })
})

describe('Binary (===|!==) type combinations:', () => {
  const valid: Array<[BinaryOperator, Value, Value]> = []
  const operators: BinaryOperator[] = ['===', '!==']
  const invalid: Array<[BinaryOperator, Value, Value]> = []

  // Every combination is valid
  operators.forEach(op => {
    mockValues.forEach(left => {
      mockValues.forEach(right => {
        valid.push([op, left, right])
      })
    })
  })

  test('Valid type combinations are OK', () => {
    valid.forEach(([operator, left, right]: [BinaryOperator, Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkBinaryExpression(context, operator, left, right)
      expect(error).toBeUndefined()
    })
  })

  test('Invalid type combinations return TypeError', () => {
    invalid.forEach(([operator, left, right]: [BinaryOperator, Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkBinaryExpression(context, operator, left, right)
      expect(error).toBeInstanceOf(rttc.TypeError)
      expect({
        operator,
        left,
        right,
        explain: error!.explain(),
        elaborate: error!.elaborate()
      }).toMatchSnapshot()
    })
  })
})

describe('Binary (<|>|<=|>=) type combinations:', () => {
  const valid: Array<[BinaryOperator, Value, Value]> = [
    ['<', num, num],
    ['<', str, str],
    ['>', num, num],
    ['>', str, str],
    ['<=', num, num],
    ['<=', str, str],
    ['>=', num, num],
    ['>=', str, str]
  ]
  const operators: BinaryOperator[] = ['<', '>', '<=', '>=']
  const invalid: Array<[BinaryOperator, Value, Value]> = []

  operators.forEach(op => {
    mockValues.forEach(left => {
      mockValues.forEach(right => {
        if (
          !valid.some(
            combination =>
              combination[0] === op && combination[1] === left && combination[2] === right
          )
        ) {
          invalid.push([op, left, right])
        }
      })
    })
  })

  test('Valid type combinations are OK', () => {
    valid.forEach(([operator, left, right]: [BinaryOperator, Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkBinaryExpression(context, operator, left, right)
      expect(error).toBeUndefined()
    })
  })

  test('Invalid type combinations return TypeError', () => {
    invalid.forEach(([operator, left, right]: [BinaryOperator, Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkBinaryExpression(context, operator, left, right)
      expect(error).toBeInstanceOf(rttc.TypeError)
      expect({
        operator,
        left,
        right,
        explain: error!.explain(),
        elaborate: error!.elaborate()
      }).toMatchSnapshot()
    })
  })
})

describe('Ternary/if test expression type combinations:', () => {
  const valid: Value[] = [bool]
  const invalid: Value[] = []

  mockValues.forEach(value => {
    if (!valid.some(combination => combination === value)) {
      invalid.push(value)
    }
  })

  test('Valid type combinations are OK', () => {
    valid.forEach((value: Value) => {
      const context = mockRuntimeContext()
      const error = rttc.checkIfStatement(context, value)
      expect(error).toBeUndefined()
    })
  })

  test('Invalid type combinations return TypeError', () => {
    invalid.forEach((value: Value) => {
      const context = mockRuntimeContext()
      const error = rttc.checkIfStatement(context, value)
      expect(error).toBeInstanceOf(rttc.TypeError)
      expect({
        value,
        explain: error!.explain(),
        elaborate: error!.elaborate()
      }).toMatchSnapshot()
    })
  })
})

describe('Member expression type combinations:', () => {
  const valid: Array<[Value, Value]> = [[obj, str], [arr, num]]
  const invalid: Array<[Value, Value]> = []

  mockValues.forEach(left => {
    mockValues.forEach(right => {
      if (!valid.some(combination => combination[0] === left && combination[1] === right)) {
        invalid.push([left, right])
      }
    })
  })

  test('Valid type combinations are OK', () => {
    valid.forEach(([left, right]: [Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkMemberAccess(context, left, right)
      expect(error).toBeUndefined()
    })
  })

  test('Invalid type combinations return TypeError', () => {
    invalid.forEach(([left, right]: [Value, Value]) => {
      const context = mockRuntimeContext()
      const error = rttc.checkMemberAccess(context, left, right)
      expect(error).toBeInstanceOf(rttc.TypeError)
      expect({
        left,
        right,
        explain: error!.explain(),
        elaborate: error!.elaborate()
      }).toMatchSnapshot()
    })
  })
})
