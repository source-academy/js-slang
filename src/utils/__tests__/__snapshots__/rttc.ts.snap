// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 1`] = `"Expected string or number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 2`] = `"Expected string on right hand side of operation, got number."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 3`] = `"Expected string on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 4`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 5`] = `"Expected string or number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 6`] = `"Expected string or number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 7`] = `"Expected string or number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 8`] = `"Expected string on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 9`] = `"Expected string on right hand side of operation, got number."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 10`] = `"Expected string or number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 11`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 12`] = `"Expected string or number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 13`] = `"Expected string or number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 14`] = `"Expected string on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 15`] = `"Expected string or number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 16`] = `"Expected string on right hand side of operation, got number."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 17`] = `"Expected string on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 18`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 19`] = `"Expected string or number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 20`] = `"Expected string or number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 21`] = `"Expected string or number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 22`] = `"Expected string on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 23`] = `"Expected string on right hand side of operation, got number."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 24`] = `"Expected string or number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 25`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 26`] = `"Expected string or number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 27`] = `"Expected string or number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (<|>|<=|>=) return TypeError 28`] = `"Expected string on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 1`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 2`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 3`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 4`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 5`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 6`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 7`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 8`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 9`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 10`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 11`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 12`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 13`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 14`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 15`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 16`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 17`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 18`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 19`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 20`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 21`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 22`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 23`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 24`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 25`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 26`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 27`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 28`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 29`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 30`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 31`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 32`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 33`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 34`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 35`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 36`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 37`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 38`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 39`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 40`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 41`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 42`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 43`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 44`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 45`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 46`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 47`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 48`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 49`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 50`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 51`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 52`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 53`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 54`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 55`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 56`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 57`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 58`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 59`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 60`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 61`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 62`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 63`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 64`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 65`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 66`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 67`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 68`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 69`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 70`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 71`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 72`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 73`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 74`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 75`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 76`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 77`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 78`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 79`] = `"Expected number on right hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 80`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 81`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 82`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 83`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 84`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 85`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 86`] = `"Expected number on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 87`] = `"Expected number on left hand side of operation, got string."`;

exports[`Invalid binary type combinations for (-|*|/|%) return TypeError 88`] = `"Expected number on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for + return TypeError 1`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for + return TypeError 2`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for + return TypeError 3`] = `"Expected string on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for + return TypeError 4`] = `"Expected string on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for + return TypeError 5`] = `"Expected string on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for + return TypeError 6`] = `"Expected string on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for + return TypeError 7`] = `"Expected number on right hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for + return TypeError 8`] = `"Expected number on right hand side of operation, got function."`;

exports[`Invalid binary type combinations for + return TypeError 9`] = `"Expected string on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for + return TypeError 10`] = `"Expected string on left hand side of operation, got boolean."`;

exports[`Invalid binary type combinations for + return TypeError 11`] = `"Expected string on left hand side of operation, got function."`;

exports[`Invalid binary type combinations for + return TypeError 12`] = `"Expected string on left hand side of operation, got function."`;

exports[`Invalid ternary/if test expressions return TypeError 1`] = `"Expected boolean as condition, got number."`;

exports[`Invalid ternary/if test expressions return TypeError 2`] = `"Expected boolean as condition, got string."`;

exports[`Invalid ternary/if test expressions return TypeError 3`] = `"Expected boolean as condition, got function."`;

exports[`Invalid ternary/if test expressions return TypeError 4`] = `"Expected boolean as condition, got number."`;

exports[`Invalid ternary/if test expressions return TypeError 5`] = `"Expected boolean as condition, got string."`;

exports[`Invalid ternary/if test expressions return TypeError 6`] = `"Expected boolean as condition, got function."`;

exports[`Invalid unary type combinations return TypeError 1`] = `"Expected boolean, got number."`;

exports[`Invalid unary type combinations return TypeError 2`] = `"Expected boolean, got string."`;

exports[`Invalid unary type combinations return TypeError 3`] = `"Expected boolean, got function."`;

exports[`Invalid unary type combinations return TypeError 4`] = `"Expected number, got boolean."`;

exports[`Invalid unary type combinations return TypeError 5`] = `"Expected number, got string."`;

exports[`Invalid unary type combinations return TypeError 6`] = `"Expected number, got function."`;

exports[`Invalid unary type combinations return TypeError 7`] = `"Expected number, got boolean."`;

exports[`Invalid unary type combinations return TypeError 8`] = `"Expected number, got string."`;

exports[`Invalid unary type combinations return TypeError 9`] = `"Expected number, got function."`;

exports[`Invalid unary type combinations return TypeError 10`] = `"Expected boolean, got number."`;

exports[`Invalid unary type combinations return TypeError 11`] = `"Expected boolean, got string."`;

exports[`Invalid unary type combinations return TypeError 12`] = `"Expected boolean, got function."`;

exports[`Invalid unary type combinations return TypeError 13`] = `"Expected number, got boolean."`;

exports[`Invalid unary type combinations return TypeError 14`] = `"Expected number, got string."`;

exports[`Invalid unary type combinations return TypeError 15`] = `"Expected number, got function."`;

exports[`Invalid unary type combinations return TypeError 16`] = `"Expected number, got boolean."`;

exports[`Invalid unary type combinations return TypeError 17`] = `"Expected number, got string."`;

exports[`Invalid unary type combinations return TypeError 18`] = `"Expected number, got function."`;
